### PR TITLE
Format files in parallel

### LINKED ISSUE #1128: Format multiple files in parallel
**Is your feature request related to a problem? Please describe.**
At work we use `treefmt`, which calls a formatter with all the matching files in the repository, i.e. we end up calling `ormolu <list of all our haskell files>`.

We have ~2.5k Haskell files, and this takes about 32 seconds.

**Describe the solution you'd like**

One obvious solution would be to just run the files in parallel. That could be as simple as replacing the `mapM` [here](https://github.com/tweag/ormolu/blob/master/app/Main.hs#L56) with `mapConcurrently`.

**Describe alternatives you've considered**

- Make ormolu faster some other way.
- Get `treefmt` to call ormolu once per file, in parallel. This would also be perfectly sensible.

### PR DESCRIPTION
**Pull request by michaelpj**
Closes #1128

There is no explicit configuration for this, but it is possible for users to override by changing the RTS options to the `ormolu` executable.

### CODE REVIEW

**Comment by mrkkrp** [comment]
Thanks! I imagine you have verified that this indeed works and results in performance benefits? We have some checks on the format of the `.cabal` file, could you fix the lint issue by running `cabal format`?

**Comment by amesgen** [comment]
> We have some checks on the format of the `.cabal` file, could you fix the lint issue by running `cabal format`?

Not directly related to this PR, but I just got reminded that we use `cabal format` is removing the comments; we could switch to eg [cabal-gild](https://github.com/tfausak/cabal-gild) which preserves this (and is in general much better than `cabal format`). Any objections/shall I create a PR switching to that?

**Review by amesgen — COMMENTED** [review]
Some comments (mostly for me to look at later):

 - https://github.com/tweag/ormolu/pull/896#issuecomment-1155403707 talks about garbled output. We should check that, and either add some synchronization or check that `-N1` produces ungarbled output and tell users to use that if they want ungarbled output.
 - There could be races here: https://github.com/tweag/ormolu/blob/7a334ac3a05f3c3bf9b4516b31b7ad4e74e5aacf/src/Ormolu/Utils/IO.hs#L53-L61
   I think we don't *need* to change sth here for correctness, but there actually might be some redundant work when multiple threads try to compute the value for the same cache key?

   We could also move away from shared `NOINLINE` ad-hoc caches, and instead compute such data upfront (see also the last paragraph in #915).

**Comment by michaelpj** [comment]
> I imagine you have verified that this indeed works and results in performance benefits?

Yes, as I mentioned [here](https://github.com/tweag/ormolu/issues/1128#issuecomment-2208491493) I got a pretty good speedup (3x) on my usecase (formatting about 2.5k files).

> We should check that, and either add some synchronization or check that -N1 produces ungarbled output and tell users to use that if they want ungarbled output.

Ah right, I have only been checking with `--mode inplace`. This is a bit awkward, since in principle we need to lock around all terminal writes. I _think_ all the writes happen in `app/Main.hs`, so perhaps we can still do this reasonably cheaply.

> I think we don't need to change sth here for correctness, but there actually might be some redundant work when multiple threads try to compute the value for the same cache key?

We could also just make this thread-safe by using an `MVar` instead of an `IORef`, right?

**Comment by amesgen** [comment]
> > We should check that, and either add some synchronization or check that -N1 produces ungarbled output and tell users to use that if they want ungarbled output.
> 
> Ah right, I have only been checking with `--mode inplace`.

Maybe we could only do parallel processing for `--mode inplace`? :thinking:  

> This is a bit awkward, since in principle we need to lock around all terminal writes. I _think_ all the writes happen in `app/Main.hs`, so perhaps we can still do this reasonably cheaply.

We have this: https://github.com/tweag/ormolu/blob/7a334ac3a05f3c3bf9b4516b31b7ad4e74e5aacf/src/Ormolu/Fixity/Internal.hs#L275-L276
But it is only emitted when `--debug` is set.

> > I think we don't need to change sth here for correctness, but there actually might be some redundant work when multiple threads try to compute the value for the same cache key?
> 
> We could also just make this thread-safe by using an `MVar` instead of an `IORef`, right?

Yes :+1:

**Comment by michaelpj** [comment]
> Maybe we could only do parallel processing for --mode inplace? 🤔

I think it would be nice if it worked for `--mode check` too. TBH, I think `--mode stdin` is a bit questionable with multiple files even without parallelism... do people use that?

**Comment by mrkkrp** [comment]
> Any objections/shall I create a PR switching to that?

Indeed, it would be nice to have a formatter for `.cabal` files that preserves comments.

**Comment by michaelpj** [comment]
- I added locking around writing to stdout and I don't get garbled output
- I made the cache thread-safe. This required changing some types which is a breaking change to the `Utils` module, but I can't see a way around it since the `IORef` is exposed in the type. My suspicion is that nobody is using `Utils` even though it's exposed on Hackage, so I guess it's a judgment call whether you're okay with this.

I've continued to allow formatting multiple files with `mode stdin`, it's just going to be undefined what order you get the output in. I guess in principle if people were using this they might care about the order, but perhaps we can just wait and see if anyone complains? Fixing it would be more work for probably no users.

**Comment by michaelpj** [comment]
Also this should probably be documented, not sure where.

**Comment by mrkkrp** [comment]
> * which is a breaking change to the `Utils` module

I think we only consider the `Ormolu` module as stable, so changes in `Ormolu.Utils.IO` should be fine.

> I've continued to allow formatting multiple files with `mode stdin`, it's just going to be undefined what order you get the output in. I guess in principle if people were using this they might care about the order, but perhaps we can just wait and see if anyone complains? Fixing it would be more work for probably no users.

I think it is totally fine. If anyone uses formatting through stdin on multiple files they can complain in the issue tracker. Not sure we even need to document that.

**Review by mrkkrp — APPROVED** [review]

**Comment by mrkkrp** [comment]
@amesgen I think this is ready to be merged. What do you think?

**Review by amesgen — APPROVED** [review]
Looks good, two minor comments

I tested this on Agda (ie running `ormolu -m check $(fd -e hs . $(nix build .#hackage.Agda --print-out-paths))`), and execution time drops from ~20s to ~7s on my machine with 8 cores :tada:

File: src/Ormolu/Utils/IO.hs [hunk 1]
```diff
@@ -48,14 +50,26 @@ findClosestFileSatisfying isRightFile rootOfSearch = liftIO $ do
...
+  -- the cache for the whole time. This means that potentially
+  -- we might end up running the same action twice if they
+  -- both start before one has populated the cache.
+  cache <- readMVar cacheVar
   case M.lookup k cache of
     Just v -> pure v
     Nothing -> do
       v <- action
-      modifyIORef' cacheRef (M.insert k v)
+      modifyMVar_ cacheVar (pure . M.insert k v)
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply src/Ormolu/Utils/IO.hs:74]: > It doesn't seem to change performance, though, so I am also fine with an MVar.

I think you are right, but if you don't mind then I plead that we keep it this way from laziness :)

>  As these caches are only used for parsed .cabal files and .ormolu files, and there usually are a lot more Haskell modules than these files, the caches should fill up pretty quickly. Did/do you observe a difference in performance for your project?

I guess my concern was that we would launch threads parsing 100 Haskell files in the same cabal file, those would all simultaneously query the cache, and so each one would end up doing the work.

However, I guess this doesn't consume any more time as far as the _user_ is concerned. Whereas caching _sequential_ re-queries does save wall-clock time.

The bad case with synchronizing the whole block would instead be if we launch threads parsing 1 Haskell file in each of 100 packages. Then they would all block on the cache... but each one in turn would still have to do its work when it gets the lock.

So I think the current situation is better.

File: app/Main.hs [hunk 2]
```diff
@@ -53,7 +59,8 @@ main = do
             ExitSuccess -> Nothing
             ExitFailure n -> Just n
       errorCodes <-
-        mapMaybe selectFailure <$> mapM (formatOne' . Just) (sort xs)
+        -- Format files in parallel, solves #1128
+        mapMaybe selectFailure <$> mapConcurrently (formatOne' . Just) (sort xs)
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply app/Main.hs:63]: I think this is probably the sensible thing to do. I'm sure it won't cost us performance and it does guard against the bad case. I'll change it.

[the remark this answers is not in the document]
Comment (michaelpj) [reply app/Main.hs]: Done.

File: ormolu.cabal [hunk 3]
```diff
@@ -153,6 +153,14 @@ executable ormolu
     ormolu,
     text >=2.1 && <3,
     th-env >=0.1.1 && <0.2,
+    unliftio >=0.2.25 && <0.2.26,
```
Comment (amesgen) [inline ormolu.cabal:156]:
```suggestion
    unliftio >=0.2.10 && <0.3,
```

**Comment by mrkkrp** [comment]
Thanks!

**Pull request merged**